packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts

```diff
@@ -297,42 +295,61 @@ export class DbSyncStakePoolProvider extends DbSyncProvider(RunnableModule) impl
     );
     // Create a lookup table with cached pools: (hashId:Cardano.StakePool)
     const rewardsHistoryKey = JSON.stringify(rewardsHistoryLimit);
-    const fromCache = Object.fromEntries(
+    const cachedPromises = Object.fromEntries(
       orderedResultHashIds.map((hashId) => [
         hashId,
-        this.#cache.getVal<Cardano.StakePool>(`${IDS_NAMESPACE}/${rewardsHistoryKey}/${hashId}`)
+        this.#cache.getVal<PromiseOrValue<Cardano.StakePool | undefined>>(
+          `${IDS_NAMESPACE}/${rewardsHistoryKey}/${hashId}`
+        )
       ])
     );
-    // Compute ids to fetch from db
-    const idsToFetch = Object.entries(fromCache)
-      .filter(([_, pool]) => pool === undefined)
-      .map(([hashId, _]) => ({ id: Number(hashId), updateId: hashIdsMap[hashId] }));
-    // Get stake pools extra information
-    const { poolRelays, poolOwners, poolRegistrations, poolRetirements, poolMetrics } = await this.queryExtraPoolsData(
-      idsToFetch,
-      sortType,
-      totalStake,
-      orderedResult,
-      useBlockfrost
-    );
-    const { results, poolsToCache } = toStakePoolResults(orderedResultHashIds, fromCache, useBlockfrost, {
-      lastEpochNo: Cardano.EpochNo(lastEpochNo),
-      nodeMetricsDependencies: {
-        optimalPoolCount,
-        totalAdaAmount: BigInt(totalStake)
-      },
-      poolAPYs,
-      poolDatas,
-      poolMetrics,
-      poolOwners,
-      poolRegistrations,
-      poolRelays,
-      poolRetirements,
-      poolRewards: poolRewards.filter(isNotNil),
-      totalCount
-    });
-    // Cache stake pools core objects
-    this.cacheStakePools(poolsToCache, rewardsHistoryKey);
+
+    const queriesAndMap = async () => {
+      const fromCache = await resolveObjectOfPromises(cachedPromises);
+      // Compute ids to fetch from db
+      const idsToFetch = Object.entries(fromCache)
+        .filter(([_, pool]) => pool === undefined)
+        .map(([hashId, _]) => ({ id: Number(hashId), updateId: hashIdsMap[hashId] }));
+      // Get stake pools extra information
+      const { poolRelays, poolOwners, poolRegistrations, poolRetirements, poolMetrics } =
+        await this.queryExtraPoolsData(idsToFetch, sortType, totalStake, orderedResult, useBlockfrost);
+
+      return toStakePoolResults(orderedResultHashIds, fromCache, useBlockfrost, {
+        lastEpochNo: Cardano.EpochNo(lastEpochNo),
+        poolAPYs,
+        poolDatas,
+        poolMetrics,
+        poolOwners,
+        poolRegistrations,
+        poolRelays,
+        poolRetirements,
+        poolRewards: poolRewards.filter(isNotNil),
+        totalCount
+      });
+    };
+
+    const resultPromise = queriesAndMap();
```

Comment:
The name of the internal func`queriesAndMap` sounds a bit confusing to me. It does many things + maps at the end.

Reply:
9d34b4639 should solve this